p2p/transport/webtransport/multiaddr.go

```diff
@@ -78,3 +75,59 @@ func addrComponentForCert(hash []byte) (ma.Multiaddr, error) {
 	}
 	return ma.NewComponent(ma.ProtocolWithCode(ma.P_CERTHASH).Name, certStr)
 }
+
+// IsWebtransportMultiaddrWithCerthash returns true if the given multiaddr is a
+// well formed webtransport multiaddr. Returns the number of certhashes found.
+func IsWebtransportMultiaddr(multiaddr ma.Multiaddr) (bool, int) {
+	const (
+		init = iota
+		foundUDP
+		foundQuicV1
+		foundWebTransport
+	)
+	state := init
+	certhashCount := 0
+
+	ma.ForEach(multiaddr, func(c ma.Component) bool {
+		if c.Protocol().Code == ma.P_QUIC_V1 && state == init {
+			state = foundUDP
+		}
+		if c.Protocol().Code == ma.P_QUIC_V1 && state == foundUDP {
+			state = foundQuicV1
+		}
+		if c.Protocol().Code == ma.P_WEBTRANSPORT && state == foundQuicV1 {
+			state = foundWebTransport
+		}
+		if c.Protocol().Code == ma.P_CERTHASH && state == foundWebTransport {
+			certhashCount++
+		}
+		return true
+	})
+	return state == foundWebTransport, certhashCount
+}
+
+// IsWebtransportMultiaddrWithCerthash returns true if the given multiaddr is a
+// well formed webtransport multiaddr with a certificate hash.
+func IsWebtransportMultiaddrWithCerthash(multiaddr ma.Multiaddr) bool {
+	ok, n := IsWebtransportMultiaddr(multiaddr)
+	return ok && n > 0
+}
+
+// CopyCerthashes copies the certificate hashes from the first multiaddr to the
+// second. If there are no multiaddrs in the first, the second is returned
+// unchanged.  Otherwise a new multiaddr is returned with certhashes appeneded
+// to it
+func CopyCerthashes(existingMultiaddrWithCerthashes, multiaddrWithoutCerthashes ma.Multiaddr) ma.Multiaddr {
```

Comment:
Would it make more sense to just have a `AddCerthashes(ma.Multiaddr) ma.Multiaddr` that adds current set of certhashes?

Reply:
It would for the other thing, but not for this. It would need a reference to the
transport which would be annoying. This is a pure function that can be used in
the AddrsFactory with no other context.